sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,672 @@
+bundled together into the same device image if their *Used* sets are identical.
+
+We must also split two kernels into different device images if they have
+different `[[sycl::reqd_sub_group_size()]]` or different
+`[[sycl::reqd_work_group_size()]]` values.  The reasoning is similar as the
+aspect case.  The JIT compiler currently raises an error if it tries to compile
+a kernel that has a required sub-group size if the size isn't supported by the
+target device.  The behavior for required work-group size is less clear.  The
+Intel implementation does not raise a JIT compilation error when compiling a
+kernel that uses an unsupported work-group size, but other backends might.
+Therefore, it seems safest to split device code based required work-group size
+also.
+
+Therefore, two kernels or exported device functions are only bundled together
+into the same device image if all of the following are true:
+
+* They share the same set of *Used* aspects,
+* They either both have no required sub-group size or both have the same
+  required sub-group size, and
+* They either both have no required work-group size or both have the same
+  required work-group size.
+
+These criteria are an additional filter applied to the device code split
+algorithm after taking into account the `-fsycl-device-code-split` command line
+option.  If the user requests `per_kernel` device code split, then each kernel
+is already in its own device image, so no further splitting is required.  If
+the user requests any other option, device code is first split according to
+that option, and then another split is performed to ensure that each device
+image contains only kernels or exported device functions that meet the criteria
+listed above.
+
+#### Create the "SYCL/image-requirements" property set
+
+The DPC++ runtime needs some way to know about the *Used* aspects, required
+sub-group size, and required work-group size of an image.  Therefore, the
+post-link tool provides this information in a new property set named
+"SYCL/image-requirements".
+
+The following table lists the properties that this set may contain and their
+types:
+
+Property Name             | Property Type
+-------------             | -------------
+"aspect"                  | `PI_PROPERTY_TYPE_BYTE_ARRAY`
+"reqd\_sub\_group\_size"  | `PI_PROPERTY_TYPE_BYTE_ARRAY`
+"reqd\_work\_group\_size" | `PI_PROPERTY_TYPE_BYTE_ARRAY`
+
+There is an "aspect" property if the image's *Used* set is not empty.  The
+value of the property is an array of `uint32` values, where each `uint32` value
+is the numerical value of an aspect from `enum class aspect`.  The size of the
+property (which is always divisible by `4`) tells the number of aspects in the
+array.
+
+There is a "reqd\_sub\_group\_size" property if the image contains any kernels
+with a required sub-group size.  The value of the property is a `uint32` value
+that tells the required size.  (The device code split algorithm ensures that
```

Comment:
Here we should also take into account `auto` and `primary` sub-group sizes from `named_sub_group_size` attribute

Reply:
Agreed.  Done in a818156a5c94c178791334b76bca83bb59d70e56